### ISSUE TITLE
Support per-lockfile (pip constraints file) Python interpreter version selection

### ISSUE THREAD

**Issue by webknjaz** [dependabot/dependabot-core#4607]
Huh?
====

Python libraries/frameworks typically support a range of interpreter versions. They are tested in different environments. The Python ecosystem is rather sophisticated — the distribution packages may declare that they support one range of Python versions but not the other. Besides, some releases may ship binary distributions (wheels) targeting different platforms (OS+arch). On top of that, there's a possibility of marking some of the dependencies as conditional (usually, depending on the Python interpreter version, OS, or architecture). This results in different dependency trees produced when they are generated under different platforms/Pythons with pip-tools.
Sometimes, the generated file may be compatible with other platforms but this is not always true. I've documented one of such cases in https://github.com/dependabot/dependabot-core/issues/4606.

This is why I've got a number of pip-tools-generated files per Python version x OS x arch that are picked up depending on the current environment.
The problem is that each of those should be generated + updated in the environment that it targets. And Dependabot is currently just using only one Python version under Linux.

I suppose that adding more OSs+arch versions into the mix may be problematic but Dependabot should at least allow declaring a per-file Python version which is what this FR is about.

I've seen some attempts to infer the interpreter version from the environment markers in `constraints.txt` files but this is not enough and that matching mechanism is imperfect. In the previous issue (https://github.com/dependabot/dependabot-core/issues/4606), what's happening is that Dependabot tries updating a constraints file for Python 2.7 while using Python 3.10 and fails on the `watchdog` package which is an indirect transitive dependency that has an upper bound set to `< 3.6` in `python_requires` (https://github.com/gorakhargosh/watchdog/blob/v0.10.7/setup.py#L159) for v0.10.7 while v1.0.0 has `>= 3.6` meaning that Python 2.7-3.5 must use `watchdog < 1` and Python 3.6-3.11 must use `watchdog >= 1` — there's not a single `watchdog` release supporting the whole Python version range of `>= 2.7, <= 3.11`.

This is not a problem for downstream projects that target a single deployment environment but is a showstopper for projects that are frameworks or libraries.

Now what
========

I don't see any possibility to declare this in `dependabot.yml` but how about having a special comment in the requirements/constraints files or maybe better have that encoded in the file name. My current approach uses the latter — the base file name format I use is `<preffix>-<python tag>-<platform>-<architecture>`:
```console
requirements/tox-build-dists-cp310-linux-x86_64.in
requirements/tox-build-dists-cp310-linux-x86_64.txt
requirements/tox-build-dists-cp39-linux-x86_64.in
requirements/tox-build-dists-cp39-linux-x86_64.txt
```

Another idea could be retrieving this information from comments that pip-tools outputs like:
```console
# This file is autogenerated by pip-compile with python 3.10
```

**Comment by jeffwidman** [comment]
Related: https://github.com/dependabot/dependabot-core/issues/4216

**Comment by jeffwidman** [comment]
Let's track this under #4216 and use the header command generated by `pip-compile` as a more deterministic way of extracting this information.

**Comment by webknjaz** [comment]
@jeffwidman that issue is about the Python interpreter version and it completely misses out the platform/arch factors. I don't see how it can be tracked there since the header comment provides no such information.

**Comment by jeffwidman** [comment]
I'm a little confused here. 

I realize the arch isn't mentioned in the pip-compile headers, but your original issue description mentions extracting the interpreter version from the comments so that's what I thought you meant. My apologies.

Dependabot at GitHub is only run on Linux, so wouldn't be able to accurately bump versions on macOS, or other non-linux platforms.

You could run Dependabot standalone yourself on whatever platform you wish using dependabot-script or another entry point.

However,  if there are multiple requirements files per-OS, but their contents don't indicate the OS/Arch, and the upstream package manager has no convention for file naming,  then I don't see how Dependabot could extract the platform requirements without adding extra conventions.

Keep in mind that Dependabot is glue code, not the package manager itself. I don't see us getting into the business of asking users to name their manifests a certain way or add extra magic marker comments for us to infer  the correct arch/platform. Instead, you should work with the upstream package manager to add that feature for all users in the ecosystem and then we will happily support it.

I'm not sure there's much we can do here automatically, although configuring only specific manifests to be monitored might be one workaround:
* https://github.com/dependabot/dependabot-core/issues/2883

**Comment by webknjaz** [comment]
Yeah, I must've emphasized just one side of the problem originally and forgot about it. But yes, if it's possible to run it in GHA, maybe that would solve the problem. Anyway, I have no problem using pip-tools directly. It's just that it's less integrated and the integrated version doesn't expose enough toggles at the moment. Ignoring certain files would probably be a stop gap. It feels like this could use a config file with proper env mapping.

**Comment by JonathanRenon-EDB** [comment]
Hello,
is there any update regarding this issue ?

**Comment by JamieMagee** [comment]
Done — `--python-version` is now supported in pip-compile (PR #13812), so you can specify the Python version per lockfile.

**Comment by webknjaz** [comment]
@sirosen something to feature in our docs?

**Comment by webknjaz** [comment]
@JamieMagee wait a second.. Does that only implement this for `uv` and not `pip-tools`?

**Comment by webknjaz** [comment]
And still no way to work per OS.

**Comment by sirosen** [comment]
Even if we wanted to imitate the uv interface, pip-tools isn't a Python version manager. If we supported `--python-version 3.12`, that would mean, for us, trying to find `python3.12` in the PATH.

So, we could match that interface (which I'm not sure we should!), but the requirement for dependabot runners to support it would be to have the target Pythons installed.

I'm slightly concerned in general that `uv pip-compile` puts indirect pressure on `pip-tools` to evolve in ways which only make sense for `uv`.

-----

I would understand if dependabot doesn't want to try to support multiplatform usage, since the only way to get true support is to run on the target platforms.

**Comment by webknjaz** [comment]
Sounds like Dependabot should really have separate entries for `pip` and `uv` instead of pretending it uses `pip` all the time, while in reality it calls a bunch of different tools, not all having standards behind..

**Comment by webknjaz** [comment]
So I can conclude #13812 does not actually solve this issue despite claiming that it does.

**Comment by JamieMagee** [comment]
@webknjaz I went through and tried to match up some old issues with PRs, but if this isn't resolved I'll reopen it.
